### DISCUSSION TITLE
Directus as a SSO provider

### DISCUSSION THREAD

**Discussion by beasteers** [directus/directus#23786] [Feature Requests]
### Summary

Add OIDC endpoints so that external services can use Directus directly as a single-sign on provider. 

### Basic Example

We would just need to implement these endpoints.
```js
// Well-Known OIDC Discovery Endpoint
router.get('/.well-known/openid-configuration', (req, res) => {
	res.json({
		issuer: OIDC_ISSUER,
		authorization_endpoint: `${OIDC_ISSUER}/auth`,
		token_endpoint: `${OIDC_ISSUER}/token`,
		introspection_endpoint: `${OIDC_ISSUER}/token/introspect`,
		userinfo_endpoint: `${OIDC_ISSUER}/userinfo`,
		end_session_endpoint: `${OIDC_ISSUER}/logout`,
		jwks_uri: `${OIDC_ISSUER}/certs`,
		...
	});
});
```
Using existing libraries, it may be even simpler. Here's an example `node-oidc-provider` configuration:
```js
import * as accountService from "../services/account.service";

export const configuration: Configuration = {
  async findAccount(ctx, id) {
    const account = { emailVerified: true, email: "ebrahimmfadae@gmail.com" };
    return (
      account && {
        accountId: id,
        async claims(_, scope) {
          if (!scope) return undefined;
          const openid = { sub: id };
          const email = {
            email: account.email,
            email_verified: account.emailVerified,
          };
          const accountInfo = {};
          if (scope.includes("openid")) Object.assign(accountInfo, openid);
          if (scope.includes("email")) Object.assign(accountInfo, email);
          return accountInfo;
        },
      }
    );
  },
  clients: [
    {
      client_id: "app",
      client_secret: "scorpion",
      redirect_uris: ["http://localhost:3005/cb"],
      grant_types: ["authorization_code"],
      scope: "openid email profile phone address offline_access",
    },
  ],
  claims: {
    address: ["address"],
    email: ["email", "email_verified"],
    phone: ["phone_number", "phone_number_verified"],
    profile: [
      "birthdate",
      "family_name",
      "gender",
      "given_name",
      "locale",
      "middle_name",
      "name",
      "nickname",
      "picture",
      "preferred_username",
      "profile",
      "updated_at",
      "website",
      "zoneinfo",
    ],
  },
};
```

### Motivation

This would remove the need for external user management, like Keycloak, Ory, etc. And would allow for easier management of users directly with directus, adding user metadata, etc.

I suggest it because if done right, it could be very low hanging fruit!

### Detailed Design

The best place for this is likely an extension.

**IMPORTANT: This is only worthwhile if the Oauth2/OIDC implementation can be offloaded to external standardized libraries, where the added functionality is just a thin adapter around Directus's storage backend.**

The best maintained library seems to be this: https://github.com/panva/node-oidc-provider

Here is an example implementation: https://dev.to/ebrahimmfadae/develop-an-openid-server-with-nodejs-typescript-9n1

Directus Changes:
 * `directus_oidc_clients`: a place to store client details

### Must Have:
- [OIDC compliance](https://auth0.com/docs/authenticate/protocols/openid-connect-protocol)
  - well-known endpoint
  - token generation: id_token, access_token, refresh_token
  - grants: implicit, authorization_code, client_credentials, password, refresh_token
- basic client configuration

### Should Have:
- basic claim mapping (directus_users/roles/permissions fields to JWT fields)
- token revocation (e.g. `directus_oidc_revoked_tokens` with some TTL cleaning strategy)

### Could Have:
- client registration
- client roles
- oauth2 resource server/grants

### Drawbacks

Directus has its own non-JWT token, we could either just have two tokens (the easier option) or migrate all of directus to JWTs.

Having two tokens would just mean that you may have to login twice.

### Alternatives

Keep using an external service like Keycloak.

### Adoption Strategy

This doesn't have to affect normal Directus login strategy.

At most, it may be beneficial for Directus to use consistent JWTs instead of opaque tokens, but this shouldn't affect anyone.

### Unresolved Questions

Will directus's roles/permissions/user-metadata be expressive enough to specify token claims needed for external services?

**Comment by robsoncombr** [comment]
+1

Hey, today I spent all time trying to bring a Ory Stack up to authenticate Directus, if we had this Directus would resolve all pains.

**Comment by ryoakirasan** [comment]
https://github.com/ryoakirasan/directus-extension-openid-connect-provide

I implemented Directus as an OIDC provider and packaged it as an extension.

**Reply by beasteers** [reply]
Amazing! 🤩 are there API endpoints so you can provision clients on first bring up?

**Reply by ryoakirasan** [reply]
> Amazing! 🤩 are there API endpoints so you can provision clients on first bring up?
API endpoints? 
This is a hybrid application where I integrated Page Module, endpoint, and hook. There's a video example in the GitHub readme. I successfully implemented login on Gitea.

After installation, you need to restart the Directus service.

**Reply by ryoakirasan** [reply]
```json
{
  "issuer": "http://host.docker.internal:8055",
  "authorization_endpoint": "http://host.docker.internal:8055/oidc-provider/authorize",
  "token_endpoint": "http://host.docker.internal:8055/oidc-provider/token",
  "userinfo_endpoint": "http://host.docker.internal:8055/oidc-provider/userinfo",
  "jwks_uri": "http://host.docker.internal:8055/oidc-provider/jwks",
  "scopes_supported": [
    "openid",
    "email",
    "profile",
    "offline_access"
  ],
  "response_types_supported": [
    "code",
    "id_token token"
  ],
  "grant_types_supported": [
    "authorization_code",
    "refresh_token"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic"
  ],
  "claims_supported": [
    "sub",
    "name",
    "email",
    "email_verified",
    "preferred_username",
    "role"
  ]
}
```
